### PR TITLE
feat: ColorPicker implement `panelRender` API

### PR DESCRIPTION
**Pull request by RedJue**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
#42749
[[RFC] ColorPicker implement panelRender API](https://github.com/ant-design/ant-design/discussions/42833)

### 💡 Background and solution
Add title
image
Horizontal layout
image

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    ColorPicker implement `panelRender` API       |
| 🇨🇳 Chinese |     颜色选择器实现 `panelRender` API      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at f75c89f</samp>

Added a `panelRender` prop to the `ColorPicker` component to enable custom panel rendering. Created new components and contexts to support the prop and its functionality. Added a new demo and updated the documentation to show the usage of the prop and its custom components. Fixed a style conflict between the `Popover` and the `RcColorPicker` components.

### <samp>🤖 Generated by Copilot at f75c89f</samp>

*  Add `panelRender` prop to `ColorPicker` component to allow customizing the color picker panel ([link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR44-R47), [link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR74), [link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-83dba3026e7dc41ad64d3a2f4d484eeb085f0c3798dc2316fbb90d23520fca6eR32))
*  Pass `panelRender` prop to `RcColorPicker` component and rename `style` prop to `overlayInnerStyle` for `Popover` component ([link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR166), [link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fL167-R173))
*  Refactor `ColorPickerPanel` component to use `PanelPicker` and `PanelPresets` components and their context providers ([link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-3abf1a34a643a7087dfbaf3039447778574de2b61e627f4c1218297fb961c70fL7-R9), [link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-3abf1a34a643a7087dfbaf3039447778574de2b61e627f4c1218297fb961c70fL19-R63))
*  Create `PanelPicker` component to render `ColorClear`, `ColorInput`, and `panel` prop from `RcColorPicker` component using `PanelPickerContext` ([link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-a04e15c33ec6c3cfe7d1f5772c30cd22e640b5f00500f1404d652d83f825e8c9R1-R39), [link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-8f679cdde821ac22f64e05be8ed68f7e0d54000bd6650ea3a6a339a10853b4e2R1-R10))
*  Create `PanelPresets` component to render `ColorPresets` component using `PanelPresetsContext` ([link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-2e31c410f79876e7d508c2f7b2e06f8129dcfe74a9c6e980bedd1e8100d57bfaR1-R22), [link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-8f679cdde821ac22f64e05be8ed68f7e0d54000bd6650ea3a6a339a10853b4e2R1-R10))
*  Move `display` and `flexDirection` styles from `${componentCls}-panel` selector to `${componentCls}-inner-panel` selector in `components/color-picker/style/index.ts` ([link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-1b0771e050d88aebe327dae413b7f31f0797bcee1515f3c6bb2cd890ba0929d0L102-R105))
*  Add test case for `panelRender` prop and its custom components in `components/color-picker/__tests__/index.test.tsx` ([link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-aa06460351e8630f305cf6509e8137984e0f167ab2560da22874033189234fb4R369-R392))
*  Add demo and documentation for `panelRender` prop and its custom components in `components/color-picker/demo/panel-render.md`, `components/color-picker/demo/panel-render.tsx`, and `components/color-picker/index.zh-CN.md` ([link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-bc8e2737396589190c3453c739f0a92dddf4066c06b05ebfd89f8a7ea3f978c4R1-R7), [link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-e00c41c079465e93600695586a06b1a5777aad2bf2ca30ea7f97c7a953a58550R1-R116), [link](https://github.com/ant-design/ant-design/pull/43134/files?diff=unified&w=0#diff-9d1ff106a032e484c8f982279c4e1295d23f69904a26e020e99310c446e93249R32))

### CODE REVIEW

File: components/color-picker/ColorPicker.tsx [hunk 1]
```diff
@@ -158,13 +163,14 @@ const ColorPicker: CompoundedComponent = (props) => {
...
     presets,
+    panelRender,
     format: formatValue,
     onFormatChange: setFormatValue,
   };
 
   return wrapSSR(
     <Popover
-      style={styles?.popup}
+      overlayInnerStyle={styles?.popup}
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:173]: 这个应该在 styles 里加，而不是直接替换。会影响现有逻辑

Comment (RedJue) [reply components/color-picker/ColorPicker.tsx:173]: style 好像没生效，代码里看了下 style 没有透传下去

**Comment by li-jia-nan** [comment]
冲突了

**Comment by RedJue** [comment]
> 冲突了
刚合了另一个 pr 我一会处理一下

File: components/color-picker/ColorPickerPanel.tsx [hunk 2]
```diff
@@ -16,43 +16,51 @@ interface ColorPickerPanelProps extends ColorPickerBaseProps {
...
-      {panel}
-      <ColorInput value={color} onChange={onChange} prefixCls={prefixCls} {...injectProps} />
-      {Array.isArray(presets) && (
-        <>
-          <Divider className={`${colorPickerPanelPrefixCls}-divider`} />
-          <ColorPresets value={color} presets={presets} prefixCls={prefixCls} onChange={onChange} />
-        </>
-      )}
-    </div>
+    <PanelPickerProvider value={{ panel, ...panelPickerProps }}>
```
Comment (MadCcc) [inline components/color-picker/ColorPickerPanel.tsx:48]: 这个应该要 memo 一下，不然每次都会是新的对象。
应该有 lint 规则来着

File: components/color-picker/index.zh-CN.md [hunk 3]
```diff
@@ -53,6 +54,7 @@ group:
 | destroyTooltipOnHide | 关闭后是否销毁弹窗 | `boolean` | false | 5.7.0 |
 | showText | 显示颜色文本 | boolean \| `(color: Color) => React.ReactNode` | - | 5.7.0 |
 | size | 设置触发器大小 | `large` \| `middle` \| `small` | `middle` | 5.7.0 |
+| panelRender | 自定义渲染面板 | `(panel: React.ReactNode, extra: { components: { Picker: FC; Presets: FC } }) => React.ReactNode` | - | 5.7.0 |
```
Comment (MadCcc) [inline components/color-picker/index.zh-CN.md:57]: api 还是按字母顺序排一下好，现在有点乱了

File: components/color-picker/ColorPickerPanel.tsx (hunk 2 above)
Comment (RedJue) [reply components/color-picker/ColorPickerPanel.tsx:48]: 每次value变了都需要重新生成的，感觉意义不大？

File: components/color-picker/ColorPickerPanel.tsx [hunk 4]
```diff
@@ -16,43 +16,51 @@ interface ColorPickerPanelProps extends ColorPickerBaseProps {
...
+  // ====================== Render ======================
+  const innerPanel = (
+    <>
+      <PanelPicker />
+      {Array.isArray(presets) && <Divider className={`${prefixCls}-inner-panel-divider`} />}
+      <PanelPresets />
+    </>
+  );
 
   const extraPanelRender = (panel: React.ReactNode) => (
```
Comment (MadCcc) [inline components/color-picker/ColorPickerPanel.tsx:47]: rc-color-picker 是不是不需要 panelRender 了？本身只有 picker 的部分，直接用就可以了

Comment (MadCcc) [reply components/color-picker/ColorPickerPanel.tsx:47]:
```
<Popover
  content= {
    <>
      <RcColorPicker />
      <Divider />
      <ColorInpur />
    </>
  }
>
  {children}
</Popover>
```
这样？

File: components/color-picker/ColorPickerPanel.tsx (hunk 2 above)
Comment (MadCcc) [reply components/color-picker/ColorPickerPanel.tsx:48]: context 一般还是处理下，避免一下无效渲染

File: components/color-picker/ColorPickerPanel.tsx (hunk 4 above)
Comment (RedJue) [reply components/color-picker/ColorPickerPanel.tsx:47]: 感觉还是需要的，不然节点没办法注入到 panel 里面。

before
```tsx
<div class="panel">
  ... // insert
</div>
```

after

``` tsx
<>
<div class="panel"></div>
...   // insert
</>
```

File: components/color-picker/ColorPickerPanel.tsx (hunk 2 above)
Comment (RedJue) [reply components/color-picker/ColorPickerPanel.tsx:48]: ok，那我也包一下

File: components/color-picker/ColorPickerPanel.tsx (hunk 4 above)
Comment (RedJue) [reply components/color-picker/ColorPickerPanel.tsx:47]: 如果不考虑整体一致性的话，这样也是可行的，这样自定义的组件就相对独立存在了和ColorPicker 本身就无关了

File: components/color-picker/ColorPickerPanel.tsx [hunk 5]
```diff
@@ -18,49 +17,63 @@ interface ColorPickerPanelProps extends ColorPickerBaseProps {
...
+  const panelPickerProps = React.useMemo(
+    () => ({
+      prefixCls,
+      value: color,
+      onChange,
+      onClear,
+      onChangeComplete,
+      ...injectProps,
+    }),
+    [prefixCls, color, onChange, injectProps],
```
Comment (MadCcc) [inline components/color-picker/ColorPickerPanel.tsx:40]: 有 injectProps 的话这个 memo 就没有意义了，可以先去掉吧

File: components/color-picker/ColorPicker.tsx [hunk 6]
```diff
@@ -43,9 +43,13 @@ export type ColorPickerProps = Omit<
...
   presets?: PresetsItem[];
   arrow?: boolean | { pointAtCenter: boolean };
+  panelRender?: (
+    panel: React.ReactNode,
+    extra: { components: { Picker: FC; Presets: FC } },
+  ) => React.ReactNode;
   showText?: boolean | ((color: Color) => React.ReactNode);
-  styles?: { popup?: CSSProperties };
   size?: SizeType;
+  styles?: { popup?: CSSProperties; overlayInnerStyle?: CSSProperties };
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:52]:
```suggestion
  styles?: { popup?: CSSProperties; popupOverlayInner?: CSSProperties };
```

**Review by MadCcc — APPROVED** [review]

**Pull request merged**